File: answers/other/answers/lintcode/Max Points on a Line.cpp
/**
 * Definition for a point.
 * struct Point {
 *     int x;
 *     int y;
 *     Point() : x(0), y(0) {}
 *     Point(int a, int b) : x(a), y(b) {}
 * };
 */
class Solution {
public:
    /**
     * @param points an array of point
     * @return an integer
     */
    int maxPoints(vector<Point>& points) {
        // Write your code here
        if (0 == points.size()) {
            return 0;
        }
        unordered_map<double, int> umap;
        int max_point = 1;
        for (size_t i = 0; i < points.size(); ++ i) {
            umap.clear();
            int same_point = 0;
            int count = 1;
            for (size_t j = i + 1; j < points.size(); ++ j) {
                double slope = numeric_limits<double>::infinity();
                if (points[i].x != points[j].x) {
                    slope = 1.0 * (points[i].y - points[j].y) / (points[i].x - points[j].x);
                }
                else if (points[i].y == points[j].y) {
                    ++ same_point;
                    continue;
                }
                if (umap.end() == umap.find(slope)) {
                    umap.insert(make_pair(slope, 1));
                }
                ++ umap[slope];
                count = max(count, umap[slope]);
            }
            max_point = max(max_point, count + same_point);
        }
        return max_point;
    }
};

File: answers/leetcode/Max Points on a Line/Max Points on a Line.py
# Definition for a point.
# class Point(object):
#     def __init__(self, a=0, b=0):
#         self.x = a
#         self.y = b

class Solution(object):
    def maxPoints(self, points):
        """
        :type points: List[Point]
        :rtype: int
        """

        if len(points) < 3:
            return len(points)
        max_points = 0
        for i in range(len(points)):
            slope_map = {}
            same_points = 1
            for j in range(len(points)):
                slope = None
                if i == j:
                    continue
                elif points[i].x == points[j].x:
                    if points[i].y == points[j].y:
                        same_points += 1
                    else:
                        if sys.maxint not in slope_map:
                            slope = sys.maxint / 10
                else:
                    slope = float(points[i].y - points[j].y) / float(points[i].x - points[j].x)
                if None != slope:
                    if slope not in slope_map:
                        slope_map[slope] = 0
                    slope_map[slope] += 1
            #print "debug", slope_map, same_points
            max_points = max(max_points, (max(slope_map.values()) if len(slope_map) else 0) + same_points)
        return max_points
